File: DoubtClass-36(MinimumEggDrops-PalindromePartioning-WordBreak)/MinimumEggDrops.java
class Solution {
	int dp[][];
	int minEggDrops(int eggs, int floors) {
		if(floors <= 1 || eggs == 1) {
			return dp[eggs][floors] = floors;
		}
		if(dp[eggs][floors] != -1) {
			return dp[eggs][floors];
		}
		int minimum = Integer.MAX_VALUE, result;
		for(int i = 1; i <= floors; i++) {
			result = 1 + Math.max(minEggDrops(eggs - 1, i - 1), minEggDrops(eggs, floors - i));
			minimum = Math.min(result, minimum);
		}
		return dp[eggs][floors] = minimum;
	}
	int minimumMoves(int k, int n) {
		dp = new int[k + 1][n + 1];
		for(int i = 0; i <= k; i++) {
			Arrays.fill(dp[i], -1);
		}
	    return minEggDrops(k, n);
	}
}

File: DoubtClass-36(MinimumEggDrops-PalindromePartioning-WordBreak)/MinimumEggDrops.cpp
int minEggDrops(int k,int n, vector<vector<int>>& dp) {
	if(n <= 1 || k == 1) {
		return n;
	}
	if(dp[k][n] != -1) {
		return dp[k][n];
	}
	int ans = INT_MAX;
	for(int i=1;i<=n;i++) {
		int result = 1 + max(minEggDrops(k-1,i-1,dp), minEggDrops(k,n-i,dp));
		ans = min(ans, result);
	}
	return dp[k][n] = ans;
}
int minimumMoves(int k, int n) {
    // add your logic here
	vector<vector<int>> dp(k+1, vector<int>(n+1,-1));
	return minEggDrops(k,n,dp);
}